pkg/controllers/management/auth/globalroles/enqueue.go

```diff
@@ -0,0 +1,144 @@
+package globalroles
+
+import (
+	"context"
+	"fmt"
+
+	mgmtv3 "github.com/rancher/rancher/pkg/generated/controllers/management.cattle.io/v3"
+	v3 "github.com/rancher/rancher/pkg/generated/norman/management.cattle.io/v3"
+	"github.com/rancher/rancher/pkg/types/config"
+	"github.com/rancher/wrangler/pkg/relatedresource"
+	"github.com/sirupsen/logrus"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+const (
+	grbGrIndex        = "mgmt-auth-grb-gr-idex"
+	grbEnqueuer       = "mgmt-auth-gr-enqueue"
+	clusterGrEnqueuer = "mgmt-auth-cluster-gr"
+	crtbGRBEnqueuer   = "mgmt-auth-crtb-grb"
+)
+
+type globalRBACEnqueuer struct {
+	grbCache      mgmtv3.GlobalRoleBindingCache
+	grCache       mgmtv3.GlobalRoleCache
+	clusterClient mgmtv3.ClusterClient
+}
+
+// runEnqueuers runs functionality that enqueues GlobalRoles/GlobalRoleBindings based on changes to the backing RBAC
+func runEnqueuers(ctx context.Context, management *config.ManagementContext) {
+	management.Wrangler.Mgmt.GlobalRoleBinding().Cache().AddIndexer(grbGrIndex, grbGrIndexer)
+	enqueuer := globalRBACEnqueuer{
+		grbCache:      management.Wrangler.Mgmt.GlobalRoleBinding().Cache(),
+		grCache:       management.Wrangler.Mgmt.GlobalRole().Cache(),
+		clusterClient: management.Wrangler.Mgmt.Cluster(),
+	}
+	relatedresource.WatchClusterScoped(ctx, grbEnqueuer, enqueuer.enqueueGRBs, management.Wrangler.Mgmt.GlobalRoleBinding(), management.Wrangler.Mgmt.GlobalRole())
+	relatedresource.WatchClusterScoped(ctx, clusterGrEnqueuer, enqueuer.clusterEnqueueGRs, management.Wrangler.Mgmt.GlobalRole(), management.Wrangler.Mgmt.Cluster())
+	relatedresource.WatchClusterScoped(ctx, crtbGRBEnqueuer, enqueuer.crtbEnqueueGRB, management.Wrangler.Mgmt.GlobalRoleBinding(), management.Wrangler.Mgmt.ClusterRoleTemplateBinding())
+}
+
+// grbGrIndexer indexes a globalRoleBinding by the globalRole it assigns to users
+func grbGrIndexer(grb *v3.GlobalRoleBinding) ([]string, error) {
+	return []string{grb.GlobalRoleName}, nil
+}
+
+// enqueueGRBs enqueues GlobalRoleBinding for a given changed GlobalRole, allowing per-cluster permissions to sync
+func (g *globalRBACEnqueuer) enqueueGRBs(namespace, name string, obj runtime.Object) ([]relatedresource.Key, error) {
+	if obj == nil {
+		return nil, nil
+	}
+	globalRole, ok := obj.(*v3.GlobalRole)
+	if !ok {
+		return nil, fmt.Errorf("unable to convert object %v to a global role", obj)
```

Comment:
Thanks for the info, I did not know about those formatting directives. I'll make an update. My plan is to add in a formatting directive for the type but also keep %v - the idea is that I would also want to see which object was passed in (specifically the name field) in case this is ever called for future debugging. But you are 100% right about the type idea, I think that's also good to include. LMK if you disagree.